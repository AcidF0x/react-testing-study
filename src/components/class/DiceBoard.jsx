import React from "react";
import Dice from "../../services/Dice";

class DiceRoller extends React.Component {
  state = {
    currentDice: null
  }

  rollDice = () => {
    const currentNumber = Dice.roll();
    this.setState({
      currentDice: currentNumber
    })
    this.props.diceRolled(currentNumber)
  }

  render() {
    return (
      <div className="dice-right-panel">
        <div className="dice-result">
          {
            this.state.currentDice !== null &&
            <h4>
              { this.state.currentDice }이/가 나왔습니다 !!
            </h4>
          }
        </div>
        <div className="dice-desc">
          <button onClick={this.rollDice}>던지기</button>
        </div>
      </div>
    )
  }
}

export default DiceRoller;
