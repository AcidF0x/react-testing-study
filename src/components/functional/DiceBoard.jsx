import React, {useState} from "react";
import Dice from "../../services/Dice";

const DiceBoard = ({diceRolled}) => {
  const [currentDice, setCurrentDice] = useState(null);
  const rollDice = () => {
    const currentNumber = Dice.roll()
    setCurrentDice(currentNumber)
    diceRolled(currentNumber)
  }

  return (
    <div className="dice-right-panel">
      <div className="dice-result">
        {
          currentDice !== null &&
          <h4>
            { currentDice }가 나왔습니다 !!
          </h4>
        }
      </div>
      <div className="dice-desc">
        <button onClick={rollDice}>던지기</button>
      </div>
    </div>
  )
}
export default DiceBoard;
