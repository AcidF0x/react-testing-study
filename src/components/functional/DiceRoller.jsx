import React, {useState} from "react";
import Dice from "../../services/Dice";
const DiceRoller = () => {
  const [totalScore, setTotalScore] = useState(0);
  const [currentDice, setCurrentDice] = useState(null);

  const resetGame = () => {
    setTotalScore(0)
    setCurrentDice(0)
  }
  const rollDice = () => {
    const currentNumber = Dice.roll()
    setCurrentDice(currentNumber)
    setTotalScore(totalScore + currentNumber)

    if (totalScore > 100) {
      resetGame()
      alert('100점이 넘어 게임이 초기화 됩니다')
    }
  }

  return (
    <>
      <h1 className="text-center">주사위 던지기</h1>
      <div className="dice-warp">
        <div className="dice-score-board">
          <div className="text-center">
            <h3>총점수</h3>
            <div className="score">
              { totalScore }
            </div>
          </div>
        </div>
        <div className="dice-right-panel">
          <div className="dice-result">
            {
              currentDice !== null &&
              <h4>
                { currentDice }가 나왔습니다 !!
              </h4>
            }
          </div>
          <div className="dice-desc">
            <button onClick={()=> rollDice()}>던지기</button>
          </div>
        </div>
      </div>
    </>
  );
}

export default DiceRoller;