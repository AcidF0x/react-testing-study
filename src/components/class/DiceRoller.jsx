import React from "react";
import DiceBoard from "./DiceBoard";
import ScoreBoard from "./ScoreBoard";

class DiceRoller extends React.Component {
  state = {
    totalScore: 0
  }
  updateTotalScore = (diceNumber) => {
    const currentScore = diceNumber + this.state.totalScore
    if (currentScore >= 100) {
      this.setState({
        totalScore: 0,
      })
      alert('100점이 넘어 게임이 초기화 됩니다')
    } else {
      this.setState({
        totalScore: currentScore,
      })
    }
  }
  render() {
    return (
      <>
        <h1 className="text-center">주사위 던지기</h1>
        <div className="dice-warp">
          <ScoreBoard totalScore={this.state.totalScore}/>
          <DiceBoard diceRolled={this.updateTotalScore}/>
        </div>
      </>
    );
  }
}
export default DiceRoller;