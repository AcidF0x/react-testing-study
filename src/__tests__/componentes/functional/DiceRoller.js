import React from "react";
import DiceRoller from "../../../components/functional/DiceRoller";
import { shallow } from "enzyme";
import Dice from "../../../services/Dice";
import DiceBoard from "../../../components/functional/DiceBoard";
window.alert = jest.fn();

describe('DiceRoller 테스트', () => {
  it('렌더링 테스트', () => {
    // Given

    // When
    const wrapper = shallow(<DiceRoller/>);

    // Then
    expect(wrapper).toMatchSnapshot();
  })
  it('총점수가 100점 미만이면 총점수를 변경한다', () => {
    // Given
    const mockDiceNumber = 5;
    const wrapper = shallow(<DiceRoller/>);
    const diceBoard = wrapper.find('DiceBoard')

    // When
    diceBoard.prop('diceRolled')(mockDiceNumber)

    // Then
    const scoreBoard = wrapper.find('ScoreBoard')
    expect(scoreBoard.props().totalScore).toBe(mockDiceNumber)
  })
  it('100점 이상 이면 점수를 0으로 만든다', () => {
    // Given
    const mockDiceNumber = 100;
    const wrapper = shallow(<DiceRoller/>);
    const diceBoard = wrapper.find('DiceBoard')

    // When
    diceBoard.prop('diceRolled')(mockDiceNumber)

    // Then
    const scoreBoard = wrapper.find('ScoreBoard')
    expect(scoreBoard.props().totalScore).toBe(0)
  })
  it('100점 이상 이면 alert 발생시킨다.', () => {
    // Given
    const mockDiceNumber = 100;
    const wrapper = shallow(<DiceRoller/>);
    const diceBoard = wrapper.find('DiceBoard');

    // When
    diceBoard.prop('diceRolled')(mockDiceNumber);

    // Then
    expect(window.alert).toHaveBeenCalledWith('100점이 넘어 게임이 초기화 됩니다');
  })
});