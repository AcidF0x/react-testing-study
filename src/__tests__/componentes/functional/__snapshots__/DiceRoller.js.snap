// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiceRoller 테스트 렌더링 테스트 1`] = `
<Fragment>
  <h1
    className="text-center"
  >
    주사위 던지기
  </h1>
  <div
    className="dice-warp"
  >
    <ScoreBoard
      totalScore={0}
    />
    <DiceBoard
      diceRolled={[Function]}
    />
  </div>
</Fragment>
`;
